### DISCUSSION TITLE
sympy / numba

### DISCUSSION THREAD

**Discussion by Peter230655** [sympy/sympy#25391] [Q&A]
I play around a lot with sympy.physics.mechanics, which invariably leads to numerically solving differentual equations.

I heared, that numba can speed up the execution, and I would like to test it.

Let's say I have a function like this:

force_vector_lam = sympy.lambdify(parameters, force_vector, cse=True)

Would this be correct:

from numba import jit

@jit 
def force_vector_num(parameters):
		return force_vector_lam(parameters)

Thanks a lot for any advice!!

**Comment by moorepants** [comment]
You should be able to use numba on any numerical function, i.e. whatever comes out of lambdify. Here is an PR that seems to show how someone else used it: https://github.com/sympy/sympy/pull/20088/files

**Comment by Peter230655** [comment]
Thanks a lot!
I will try it as soon as I am back home from my last business trip and can use my desk top again.
I will report what I will find out.

**Comment by brocksam** [comment]
My experience with Numba and `lambdify` has been that the Numba compilation times increase worse-than-linearly with the number of expressions in the lambdified function. So if you're trying to JIT a lambdified expression with a large number of expressions the compilation can take a while. But it does work well and results in much faster code than the Python code generated by vanilla `lambdify`. Worth it if you're going to be evaluating your callable many times.

I'd also suggest using `numba.njit` or `numba.jit(nopython=True)` rather than `numba.jit` (`numba.njit` and `numba.jit(nopython=True)` are equivalent). This ensures that Numba uses it's "nopython" mode (and doesn't fall back to "object" mode), which produces much faster numeric code.

**Comment by Peter230655** [comment]
Thanks for the hint!
Would 'njit' also speed up scipy functions, say if I do like this:

'@njit
def loesung(func, x0, parameters);
  return fsolve(func, x0, parameters)

Thanks again!

**Comment by brocksam** [comment]
No, Numba would fail in that case because SciPy's `fsolve` is wrapping the Fortran library MINPACK, which is already compiled and highly optimised. As a rule of thumb Numba is best left for your own pure-Python functions.

**Comment by Peter230655** [comment]
Thanks!!
You guys know so many things!
I was too old when I got into python, etc! :-(

**Comment by asmeurer** [comment]
Numba is very strict, especially when using `@njit`, so the best thing to do if you want to know if it can work with something is to just try it. If it works, that means it can speed it up, and if it gives an error, that means it doesn't support it yet.

**Reply by Peter230655** [reply]
Thanks!!
If F(..) is some, say, scipy function, I can simply put
ergebnis = njit(F(..))
and if it does not give an error, fine - if it gives an error, too bad?
What I mean: it will not run without error but give a wrong result?
Thanks again!

**Comment by asmeurer** [comment]
> @jit
def force_vector_num(parameters):
return force_vector_lam(parameters)

By the way, you don't need the level of indirection here. Putting `@njit` on top of a function definition is the same as doing `njit(function)`. So you can just do 

```py
force_vector_lam = njit(sympy.lambdify(parameters, force_vector, cse=True))
```

**Reply by Peter230655** [reply]
Thanks!!
